pyomo/contrib/gdpopt/GDPopt.py

```diff
@@ -0,0 +1,1467 @@
+        if subprob_terminate_cond is tc.optimal:
+            m.solutions.load_from(results)
+            process_feasible_solution()
+            solnFeasible = True
+        elif subprob_terminate_cond is tc.infeasible:
+            # TODO try something else? Reinitialize with different initial
+            # value?
+            logger.info('NLP subproblem was locally infeasible.')
+            # load the solution and suppress the warning message by setting
+            # solver status to ok.
+            results.solver.status = SolverStatus.ok
+            m.solutions.load_from(results)
+            self._copy_values(m, solve_data.working_model)
+            self._copy_dual_suffixes(m, solve_data.working_model,
+                                     from_map=obj_to_cuid)
+            if solve_data.current_strategy == 'PSC':
+                logger.info('Adding PSC feasibility cut.')
+                self._add_psc_cut(solve_data, nlp_feasible=False)
+            elif solve_data.current_strategy == 'LGBD':
+                logger.info('Adding GBD feasibility cut.')
+                self._add_gbd_cut(solve_data, nlp_feasible=False)
+            # Add an integer cut to exclude this discrete option
+            self._add_int_cut(solve_data)
+        elif subprob_terminate_cond is tc.maxIterations:
+            # TODO try something else? Reinitialize with different initial
+            # value?
+            logger.info('NLP subproblem failed to converge '
+                        'within iteration limit.')
+            results.solver.status = SolverStatus.ok
+            m.solutions.load_from(results)
+            if self._is_feasible(m):
+                logger.info('NLP solution is still feasible. '
+                            'Using potentially suboptimal feasible solution.')
+                process_feasible_solution()
+                solnFeasible = True
+            else:
+                # Add an integer cut to exclude this discrete option
+                self._add_int_cut(solve_data)
+        else:
+            raise ValueError(
+                'GDPopt unable to handle NLP subproblem termination '
+                'condition of %s. Results: %s'
+                % (subprob_terminate_cond, results))
+
+        if GDPopt.objective.sense == minimize:
+            solve_data.UB_progress.append(solve_data.UB)
+        else:
+            solve_data.LB_progress.append(solve_data.LB)
+
+        # Call the NLP post-solve callback
+        solve_data.subproblem_postsolve(m, solve_data)
+        return solnFeasible
+
+    def _is_feasible(self, m, constr_tol=1E-6, var_tol=1E-8):
+        for constr in m.component_data_objects(
+                ctype=Constraint, active=True, descend_into=(Block, Disjunct)):
+            # constraint is an equality
+            if constr.equality:
+                if fabs(value(constr.lower) -
+                        value(constr.body)) >= constr_tol:
+                    logger.info('%s: %s ≠ %s' % (
```

Comment:
It would be significantly better if you did not include unicode characters (not equal to) in the code base.

Reply:
Ruins the one-character symmetry with < and > to use != or the like, though

Reply:
You could always use `!` if you really want a single-character operator.